File: platform/pkg/observability/tracing/propogation.go
// platform/pkg/observability/tracing/propagation.go
package tracing

import (
	"context"
	"net/http"
)

// InjectHTTP injects the request ID into HTTP headers.
// Used by HTTP clients to propagate request IDs.
func InjectHTTP(header http.Header, requestID RequestID) {
	if requestID == "" {
		return
	}
	header.Set(RequestIDHeader, requestID.String())
}

// ExtractHTTP extracts the request ID from HTTP headers.
// Used by HTTP servers to receive request IDs.
func ExtractHTTP(header http.Header) RequestID {
	value := header.Get(RequestIDHeader)
	if value == "" {
		return ""
	}
	return ParseRequestID(value)
}

// HTTPMiddleware is HTTP middleware that extracts or generates a request ID
// and attaches it to the request context.
func HTTPMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Extract from header or generate new
		rid := ExtractHTTP(r.Header)
		if rid == "" {
			rid = NewRequestID()
		}

		// Attach to context
		ctx := WithRequestID(r.Context(), rid)
		r = r.WithContext(ctx)

		// Add to response headers for clients
		w.Header().Set(RequestIDHeader, rid.String())

		next.ServeHTTP(w, r)
	})
}

// InjectEventMetadata injects the request ID into event metadata.
// Used when publishing events to NATS/RabbitMQ.
func InjectEventMetadata(metadata map[string]string, requestID RequestID) {
	if requestID == "" {
		return
	}
	if metadata == nil {
		return
	}
	metadata[RequestIDHeader] = requestID.String()
}

// ExtractEventMetadata extracts the request ID from event metadata.
// Used when consuming events from NATS/RabbitMQ.
func ExtractEventMetadata(metadata map[string]string) RequestID {
	if metadata == nil {
		return ""
	}
	value, ok := metadata[RequestIDHeader]
	if !ok {
		return ""
	}
	return ParseRequestID(value)
}

// PropagateToEvent extracts the request ID from context and injects it into event metadata.
// Returns the request ID for convenience.
func PropagateToEvent(ctx context.Context, metadata map[string]string) RequestID {
	rid := GetRequestID(ctx)
	if rid == "" {
		rid = NewRequestID()
	}
	InjectEventMetadata(metadata, rid)
	return rid
}

// ExtractFromEvent extracts the request ID from event metadata and attaches it to context.
func ExtractFromEvent(ctx context.Context, metadata map[string]string) context.Context {
	rid := ExtractEventMetadata(metadata)
	if rid == "" {
		rid = NewRequestID()
	}
	return WithRequestID(ctx, rid)
}

File: platform/pylib/scout_common/observability/tracing/propogation.py
# platform/pylib/scout_common/observability/tracing/propagation.py
"""Propagation helpers for request IDs across different transports."""

from typing import Dict, Optional, Callable, Awaitable

from .request_id import RequestID, REQUEST_ID_HEADER
from .context import get_request_id, set_request_id


def inject_http(headers: Dict[str, str], request_id: Optional[RequestID] = None) -> None:
    """Inject request ID into HTTP headers.
    
    Used by HTTP clients to propagate request IDs.
    
    Args:
        headers: HTTP headers dictionary
        request_id: Request ID to inject (or None to use from context)
    """
    if request_id is None:
        request_id = get_request_id()
    
    if request_id is None:
        return
    
    headers[REQUEST_ID_HEADER] = str(request_id)


def extract_http(headers: Dict[str, str]) -> Optional[RequestID]:
    """Extract request ID from HTTP headers.
    
    Used by HTTP servers to receive request IDs.
    
    Args:
        headers: HTTP headers dictionary
        
    Returns:
        Request ID if found, None otherwise
    """
    # Try case-insensitive lookup
    for key in [REQUEST_ID_HEADER, REQUEST_ID_HEADER.lower()]:
        value = headers.get(key)
        if value:
            return RequestID.parse(value)
    
    return None


def inject_event_metadata(
    metadata: Dict[str, str],
    request_id: Optional[RequestID] = None
) -> None:
    """Inject request ID into event metadata.
    
    Used when publishing events to NATS/RabbitMQ.
    
    Args:
        metadata: Event metadata dictionary
        request_id: Request ID to inject (or None to use from context)
    """
    if request_id is None:
        request_id = get_request_id()
    
    if request_id is None:
        return
    
    metadata[REQUEST_ID_HEADER] = str(request_id)


def extract_event_metadata(metadata: Dict[str, str]) -> Optional[RequestID]:
    """Extract request ID from event metadata.
    
    Used when consuming events from NATS/RabbitMQ.
    
    Args:
        metadata: Event metadata dictionary
        
    Returns:
        Request ID if found, None otherwise
    """
    value = metadata.get(REQUEST_ID_HEADER)
    if value:
        return RequestID.parse(value)
    
    return None


def propagate_to_event(metadata: Dict[str, str]) -> RequestID:
    """Extract request ID from context and inject into event metadata.
    
    Args:
        metadata: Event metadata dictionary
        
    Returns:
        Request ID (from context or newly generated)
    """
    request_id = get_request_id()
    if request_id is None:
        request_id = RequestID.generate()
        set_request_id(request_id)
    
    inject_event_metadata(metadata, request_id)
    return request_id


def extract_from_event(metadata: Dict[str, str]) -> RequestID:
    """Extract request ID from event metadata and set in context.
    
    Args:
        metadata: Event metadata dictionary
        
    Returns:
        Request ID (extracted or newly generated)
    """
    request_id = extract_event_metadata(metadata)
    if request_id is None:
        request_id = RequestID.generate()
    
    set_request_id(request_id)
    return request_id


# HTTP Middleware helper
async def http_middleware(
    request,
    call_next: Callable[[any], Awaitable[any]]
):
    """HTTP middleware for FastAPI/Starlette.
    
    Extracts or generates request ID and adds to context.
    
    Args:
        request: HTTP request
        call_next: Next middleware/handler
        
    Returns:
        HTTP response with request ID header
    """
    # Extract from headers or generate
    request_id = extract_http(dict(request.headers))
    if request_id is None:
        request_id = RequestID.generate()
    
    # Set in context
    set_request_id(request_id)
    
    # Process request
    response = await call_next(request)
    
    # Add to response headers
    response.headers[REQUEST_ID_HEADER] = str(request_id)
    
    return response